fe/topology.py

```diff
@@ -390,6 +555,163 @@ def parameterize_improper_torsion(self, ff_params):
         return self._parameterize_bonded_term(ff_params, self.ff.it_handle, potentials.PeriodicTorsion)
 
 
+class DualTopologyRHFE(DualTopology):
+    """
+    Utility class used for relative hydration free energies. Ligand B is decoupled as lambda goes
+    from 0 to 1, while ligand A is fully coupled. At the same time, at lambda=0, ligand B and ligand A
+    have their charges and epsilons reduced by half.
+    """
+    def parameterize_nonbonded(self,
+        ff_q_params,
+        ff_lj_params):
+
+        qlj_params, nb_potential = super().parameterize_nonbonded(ff_q_params, ff_lj_params)
+
+        # halve the strength of the charge and the epsilon parameters
+        src_qlj_params = jax.ops.index_update(qlj_params, jax.ops.index[:, 0], qlj_params[:, 0]*0.5)
+        src_qlj_params = jax.ops.index_update(src_qlj_params, jax.ops.index[:, 2], qlj_params[:, 2]*0.5)
+        dst_qlj_params = qlj_params
+        combined_qlj_params = jnp.concatenate([src_qlj_params, dst_qlj_params])
+
+        combined_lambda_plane_idxs = np.zeros(
+            self.mol_a.GetNumAtoms() + self.mol_b.GetNumAtoms(),
+            dtype=np.int32
+        )
+        combined_lambda_offset_idxs = np.concatenate([
+            np.zeros(self.mol_a.GetNumAtoms(), dtype=np.int32),
+            np.ones(self.mol_b.GetNumAtoms(), dtype=np.int32)
+        ])
+
+        nb_potential.set_lambda_plane_idxs(combined_lambda_plane_idxs)
+        nb_potential.set_lambda_offset_idxs(combined_lambda_offset_idxs)
+
+        return combined_qlj_params, nb_potential.interpolate()
+
+
+def get_ring_membership(mol):
+    """
+    Get the membership of each atom in the mol. The algorithm
+    finds islands that are formed connected by bond bridges.
+
+    Let N = mol.GetNumAtoms()
+
+    The returned array is of length N, where each atom is marked
+    by a membership class K such that 0 < K < N.
+    """
+
+    g = networkx.Graph()
+    for bond in mol.GetBonds():
+        g.add_edge(bond.GetBeginAtomIdx(), bond.GetEndAtomIdx())
+
+    # find bridges and remove them to form islands
+    for bridge in networkx.bridges(g):
+        g.remove_edge(*bridge)
+
+    # find connected components after bridge removal
+    membership = np.zeros(mol.GetNumAtoms(), dtype=np.int32)
+    for group_idx, atom_list in enumerate(networkx.connected_components(g)):
+        for atom_idx in atom_list:
+            membership[atom_idx] = group_idx
+
+    return membership
```

Comment:
Nice! Documentation request: add comment about when to use this function vs. rdkit ringinfo https://www.rdkit.org/docs/GettingStartedInPython.html#ring-information ?

Reply:
rdkit RingInfo does something totally different. It is populated by either FastFindRings or SSSRs (ring basis). Neither is quite what we want.